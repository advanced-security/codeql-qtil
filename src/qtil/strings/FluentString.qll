bindingset[result]
bindingset[s]
FluentString fluentString(string s) { result = s }

class FluentString extends string {
  bindingset[this]
  FluentString() { any() }

  bindingset[s, this]
  FluentString append(string s) { result = this + s }

  bindingset[s, this]
  FluentString prepend(string s) { result = s + this }
}
