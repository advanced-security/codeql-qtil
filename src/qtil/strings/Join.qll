/**
 * A module for joining strings with a separator, for example, `join(":", "a", "b") = "a:b"`.
 *
 * This module is not designed to replace the `concat` aggregation in CodeQL, but may be useful in
 * rare cases where joining is required without aggregation.
 *
 * There are join predicate for joining 2 to 8 parts, all named `join`. The first argument is always
 * the separator string, and the remaining arguments are the strings to be joined.
 */

/**
 * Use the first argument as a separator to join two strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2]
string join(string sep, string v1, string v2) { result = v1 + sep + v2 }

/**
 * Use the first argument as a separator to join three strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3]
string join(string sep, string v1, string v2, string v3) { result = v1 + sep + v2 + sep + v3 }

/**
 * Use the first argument as a separator to join four strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3, v4]
string join(string sep, string v1, string v2, string v3, string v4) {
  result = v1 + sep + v2 + sep + v3 + sep + v4
}

/**
 * Use the first argument as a separator to join five strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3, v4, v5]
string join(string sep, string v1, string v2, string v3, string v4, string v5) {
  result = v1 + sep + v2 + sep + v3 + sep + v4 + sep + v5
}

/**
 * Use the first argument as a separator to join six strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3, v4, v5, v6]
string join(string sep, string v1, string v2, string v3, string v4, string v5, string v6) {
  result = v1 + sep + v2 + sep + v3 + sep + v4 + sep + v5 + sep + v6
}

/**
 * Use the first argument as a separator to join seven strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3, v4, v5, v6, v7]
string join(string sep, string v1, string v2, string v3, string v4, string v5, string v6, string v7) {
  result = v1 + sep + v2 + sep + v3 + sep + v4 + sep + v5 + sep + v6 + sep + v7
}

/**
 * Use the first argument as a separator to join eight strings.
 *
 * This predicate is overloaded to allow for 2 to 8 non-separator arguments.
 */
bindingset[sep, v1, v2, v3, v4, v5, v6, v7, v8]
string join(
  string sep, string v1, string v2, string v3, string v4, string v5, string v6, string v7, string v8
) {
  result = v1 + sep + v2 + sep + v3 + sep + v4 + sep + v5 + sep + v6 + sep + v7 + sep + v8
}
