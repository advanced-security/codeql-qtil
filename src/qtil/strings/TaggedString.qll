private import qtil.parameterization.SignatureTypes
private import qtil.inheritance.UnderlyingString
private import qtil.parameterization.Finalize

/**
 * A module for creating strings with an enum-like tag value.
 *
 * To use this module, create a type which is uniquely identified by its `toString()` member
 * predicate. Then you may refer to a tagged string by type `Tagged<YourTag>::String`.
 *
 * See `Tagged::String` for more details.
 */
module Tagged<FiniteStringableType Tag> {
  /**
   * A class for representing strings with an enum-like tag value.
   *
   * To use this class, define a type which is uniquely identified by its `toString()` member
   * predicate. Then refer to this class as `Tagged<YourTag>::String`.
   *
   * REQUIREMENTS: The toString() method of a tag must return non-overlapping results; no toString()
   * may begin with or equal the toString() result from another instance (for instance, `"foo"` and
   * `"foobar"` would be considered overlapping). The Tag type must be finite (not have
   * `bindingset[this]`) and the toString() member predicate must be reversable (not have
   * `bindingset[this]`).
   *
   * Tagged strings may be "constructed" by calling `s.make(tag, str)`, or with
   * `s.make(str).getTag() = ...`.
   *
   * Example usage:
   * ```ql
   * newtype TMyEnum = TOptionA() or TOptionB();
   * class MyEnum extends TMyEnum {
   *   string toString() {
   *     this = TOptionA() and result = "option_a"
   *     or
   *     this = TOptionB() and result = "option_b"
   *   }
   * }
   *
   * Tagged<MyEnum>::String getTaggedString() {
   *   result.make(TOptionA(), "string value")
   * }
   * ```
   */
  bindingset[this]
  class String extends Final<UnderlyingString>::Type {
    Tag tag;
    string strVal;

    String() { this = tag.toString() + strVal }

    /**
     * Holds if this tagged string has the given tag and string, which uniquely identifies and
     * therefore may "make" this instance.
     */
    bindingset[mstrVal]
    bindingset[this]
    predicate make(Tag mtag, string mstrVal) { this = mtag.toString() + mstrVal }

    /**
     * Holds if the tagged string has the given string contents, and any tag.
     *
     * To also set the tag of this string, either use `make(tag, str)`, or constrain the tag of
     * the result of this predicate with `make(str).getTag() = ...`.
     */
    bindingset[mstrVal]
    bindingset[result]
    String make(string mstrVal) {
      make(_, mstrVal) and
      result = this
    }

    /**
     * Get the tag of this tagged string.
     */
    bindingset[this]
    Tag getTag() { result = tag }

    /**
     * Holds if this tagged string has the given tag.
     */
    bindingset[this]
    predicate isTagged(Tag mtag) { tag = mtag }

    /**
     * Get the string contents of this tagged string.
     */
    bindingset[this]
    string getStr() { result = strVal }
  }
}
