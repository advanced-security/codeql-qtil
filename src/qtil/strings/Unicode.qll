private import codeql.util.Numbers

/**
 * Returns the Unicode version used by the current QL environment.
 */
extensible predicate unicodeVersion(string version);

/**
 * Provieds properties of a Unicode code point, where the property is of 'enumeration', 'catalog',
 * or 'string-valued' type.
 *
 * For example, `Block` is an enumeration property, `Line_Break` is a catalog property, and
 * `Uppercase_Mapping` is a string-valued property.
 *
 * For boolean properties, see `unicodeHasBooleanProperty`, and for numeric properties, see
 * `unicodeHasNumericProperty`.
 */
extensible predicate unicodeHasProperty(int codePoint, string propertyName, string propertyValue);

/**
 * Holds when the Unicode code point's boolean property of the given name is true.
 *
 * For example, `Alphabetic` is a boolean property that can be true or false for a code point.
 *
 * For other types of properties, see `unicodeHasProperty`.
 */
extensible predicate unicodeHasBooleanProperty(int codePoint, string propertyName);

/**
 * Provides the numeric value of a Unicode code point's numeric property.
 *
 * For example, `Numeric_Value` is a numeric property that can be an integer or a decimal value
 * for a code point.
 *
 * For other types of properties, see `unicodeHasProperty` and `unicodeHasBooleanProperty`.
 */
extensible predicate unicodeHasNumericProperty(
  int codePoint, string propertyName, float numericValue
);

/**
 * Holds when the given string contains a Unicode escape sequence (e.g. "\u0041").
 */
bindingset[input]
predicate containsUnicodeEscape(string input) { input.regexpMatch(".*\\\\u[0-9a-fA-F]{4}") }

/**
 * Holds when the given string contains only ASCII characters (code points 0-127).
 *
 * Note that if the string contains Unicode escape sequences, those are treated as multiple ASCII
 * characters, not as a single Unicode character. See `unescapeUnicode` and `containsUnicodeEscape`
 * for handling such Unicode escape sequences.
 */
bindingset[input]
predicate isAscii(string input) { input.regexpMatch("^[\\x{0000}-\\x{007F}]*$") }

/**
 * Takes a string containing Unicode escape sequences (e.g. "\u0041"), and returns the unescaped
 * string.
 */
bindingset[input]
string unescapeUnicode(string input) {
  // The process of splitting the input string and re-assembling it is not very efficient, so we
  // only do it if we know there are Unicode escapes to process.
  if containsUnicodeEscape(input)
  then
    result =
      concat(int code, int idx |
        code = unescapedCodePointAt(input, idx)
      |
        code.toUnicode() order by idx
      )
  else result = input
}

/**
 * Takes a string containing Unicode escape sequences (e.g. "\u0041"), and an offset into that
 * string, and returns the unescaped character at that offset.
 *
 * Not intended for general use; but rather as a helper for `unescapeUnicode`.
 *
 * Note that the offset is based on the output string, not the input string, as the regexp used in
 * this predicate splits the input into pieces that are either a single character or a Unicode
 * escape sequence.
 */
bindingset[input]
private int unescapedCodePointAt(string input, int index) {
  exists(string match |
    match = input.regexpFind("(\\\\u[0-9a-fA-F]{4}|.)", index, _) and
    if match.matches("\\u%")
    then result = parseHexInt(match.substring(2, match.length()))
    else result = match.codePointAt(0)
  )
}

/**
 * Holds when the given string is a not a valid UAX44 identifier.
 *
 * Essentially, a class of characters may begin or continue a valid identifier, while another class
 * of characters may only continue a valid identifier, and other characters may not appear at all.
 * See the unicode standard Annex #44 for full details.
 *
 * The second parameter `index` holds for the index of each invalid codepoint in the string.
 */
bindingset[id]
predicate nonUax44IdentifierCodepoint(string id, int index) {
  exists(int codePoint |
    codePoint = id.codePointAt(index) and
    (
      not unicodeHasBooleanProperty(codePoint, "XID_Start") and
      not unicodeHasBooleanProperty(codePoint, "XID_Continue")
      or
      index = 0 and
      not unicodeHasBooleanProperty(codePoint, "XID_Start")
    )
  )
}

/**
 * An enumeration of possible results for an NFC normalization check.
 *
 * See the unicode standard Annex #15 for full details.
 *
 * The full algorithm to detect NFC normalization is too complex to be worth implementing in
 * CodeQL, and there are currently no plans to add support for this in the CodeQL language itself.
 * However, the unicode standard does define an "NFC_Quick_Check" property which can be used to
 * quickly determine if a string is definitely normalized, definitely not normalized, or may not be
 * normalized. This type is used to represent those three possible results.
 */
newtype NormalizationCheckResult =
  DefinitelyNormalized() or
  DefinitelyNotNormalized() or
  MaybeNormalized()

/**
 * Holds when the given string is not normalized, or may not be normalized, in NFC form. This is
 * only an approximation of the full NFC normalization check process.
 *
 * See the unicode standard Annex #15 for full details.
 *
 * If all codepoints in the string are definitely normalized, then this predicate returns
 * `DefinitelyNormalized()` and an index of -1. If any codepoint is definitely not normalized, then
 * this predicate returns `DefinitelyNotNormalized()` and holds for each definitely not normalized
 * codepoint's index. Otherwise, if any codepoint may not be normalized, then this predicate returns
 * `MaybeNormalized()` and holds for each maybe not normalized codepoint's index.
 *
 * The exact algorithm to detect NFC normalization is too complex to be worth implementing in
 * CodeQL, and there are currently no plans to add support for this in the CodeQL language itself.
 * However, the unicode standard does define an "NFC_Quick_Check" property which can be used to
 * quickly determine if a string is definitely normalized, definitely not normalized, or may not be
 * normalized. This member predicate exposes that information for approximate checking of NFC
 * normalization.
 */
bindingset[id]
NormalizationCheckResult hasNfcNormalizationCheckResult(string id, int index) {
  // The process of splitting the input string and checking each codepoint is not very efficient, so we
  // only do it if we know there are non-ASCII characters to process. If the string is ASCII, then it is
  // definitely normalized.
  if isAscii(id)
  then result = DefinitelyNormalized() and index = -1
  else
    if nfcQuickCheckCodepointImpl(id, _) = DefinitelyNotNormalized()
    then
      result = DefinitelyNotNormalized() and
      nfcQuickCheckCodepointImpl(id, index) = DefinitelyNotNormalized()
    else
      if nfcQuickCheckCodepointImpl(id, _) = MaybeNormalized()
      then
        result = MaybeNormalized() and
        nfcQuickCheckCodepointImpl(id, index) = MaybeNormalized()
      else (
        result = DefinitelyNormalized() and index = -1
      )
}

/**
 * Helper predicate for `hasNfcNormalizationCheckResult` that checks the `NFC_QC` property of each
 * codepoint in the string.
 *
 * The result is _only_ `DefinitelyNotNormalized()` or `MaybeNormalized()`. If all codepoints are
 * definitely normalized, then this predicate does not hold at all.
 */
bindingset[id]
private NormalizationCheckResult nfcQuickCheckCodepointImpl(string id, int index) {
  exists(int codePoint, string noOrMaybe |
    codePoint = id.codePointAt(index) and
    unicodeHasProperty(codePoint, "NFC_QC", noOrMaybe) and
    noOrMaybe = ["N", "M"] and
    result = quickCheckResultFromString(noOrMaybe)
  )
}

/**
 * Translates an `NFC_QC` property value string into a `NormalizationCheckResult`.
 */
private NormalizationCheckResult quickCheckResultFromString(string noOrMaybe) {
  noOrMaybe = "Y" and result = DefinitelyNormalized()
  or
  noOrMaybe = "N" and result = DefinitelyNotNormalized()
  or
  noOrMaybe = "M" and result = MaybeNormalized()
}
