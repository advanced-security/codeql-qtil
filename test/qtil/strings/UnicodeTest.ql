import qtil.testing.Qnit
import qtil.strings.Unicode

class UnicodeHasPropertyTest extends Test, Case {
  override predicate run(Qnit test) {
    if unicodeHasProperty("A".codePointAt(0), "General_Category", "Lu")
    then test.pass("Unicode is available and has General_Category 'Lu' for 'A'.")
    else test.fail("Unicode does not have General_Category 'Lu' for 'A'.")
  }
}

class UnicodeHasBooleanPropertyTest extends Test, Case {
  override predicate run(Qnit test) {
    if unicodeHasBooleanProperty(65, "Alphabetic")
    then test.pass("Unicode is available and 'A' is alphabetic.")
    else test.fail("Unicode does not have 'A' as alphabetic.")
  }
}

class UnicodeHasNumericPropertyTest extends Test, Case {
  override predicate run(Qnit test) {
    if unicodeHasNumericProperty("9".codePointAt(0), "Numeric_Value", 9)
    then test.pass("Unicode is available and '9' has numeric value 9.")
    else test.fail("Unicode does not have '9' with numeric value 9.")
  }
}

class StringContainsUnicodeEscapeTest extends Test, Case {
  override predicate run(Qnit test) {
    if containsUnicodeEscape("Hello \\u0041")
    then test.pass("String contains Unicode escape sequence.")
    else test.fail("String does not contain Unicode escape sequence.")
  }
}

class StringDoesNotContainUnicodeEscapeTest extends Test, Case {
  override predicate run(Qnit test) {
    if not containsUnicodeEscape("Hello A")
    then test.pass("String does not contain Unicode escape sequence.")
    else test.fail("String contains Unicode escape sequence.")
  }
}

class StringIsAsciiTest extends Test, Case {
  override predicate run(Qnit test) {
    if isAscii("Hello!") then test.pass("String is ASCII.") else test.fail("String is not ASCII.")
  }
}

class StringIsNotAsciiTest extends Test, Case {
  override predicate run(Qnit test) {
    if not isAscii("Hello ñ")
    then test.pass("String is not ASCII.")
    else test.fail("String is ASCII.")
  }
}

class UnescapeUnicodeTest extends Test, Case {
  override predicate run(Qnit test) {
    if unescapeUnicode("Hello \\u0041") = "Hello A"
    then test.pass("Unicode escape sequence unescaped correctly.")
    else test.fail("Unicode escape sequence not unescaped correctly.")
  }
}

class UnescapeUnicodeNoEscapeTest extends Test, Case {
  override predicate run(Qnit test) {
    if unescapeUnicode("Hello A") = "Hello A"
    then test.pass("String without escape sequence remains unchanged.")
    else test.fail("String without escape sequence changed incorrectly.")
  }
}

class UnescapeUnicodeMultipleEscapesTest extends Test, Case {
  override predicate run(Qnit test) {
    if unescapeUnicode("Hello \\u0041 and \\u00F1") = "Hello A and ñ"
    then test.pass("Multiple Unicode escape sequences unescaped correctly.")
    else test.fail("Multiple Unicode escape sequences not unescaped correctly.")
  }
}

class NonUax44IdentifierCodepointAsciiTest extends Test, Case {
  override predicate run(Qnit test) {
    if not nonUax44IdentifierCodepoint("asciiIdentifier", _)
    then test.pass("ASCII characters are valid UAX44 identifier codepoints.")
    else test.fail("ASCII characters are not recognized as valid UAX44 identifier codepoints.")
  }
}

class NonUax44IdentifierCodepointInvalidStartTest extends Test, Case {
  override predicate run(Qnit test) {
    if nonUax44IdentifierCodepoint("1invalid", 0)
    then test.pass("'1' is an invalid start for a UAX44 identifier.")
    else test.fail("'1' is not recognized as an invalid start for a UAX44 identifier.")
  }
}

class NonUax44IdentifierCodepointValidUnicodeCharsTest extends Test, Case {
  override predicate run(Qnit test) {
    if not nonUax44IdentifierCodepoint("valid_ñ_id", _)
    then test.pass("Valid Unicode characters are accepted in UAX44 identifiers.")
    else test.fail("Valid Unicode characters are not recognized as valid in UAX44 identifiers.")
  }
}

class NonUax44IdentifierCodepointInvalidContinueTest extends Test, Case {
  override predicate run(Qnit test) {
    if nonUax44IdentifierCodepoint("valid_id!", 8)
    then test.pass("'!' is an invalid continuation for a UAX44 identifier.")
    else test.fail("'!' is not recognized as an invalid continuation for a UAX44 identifier.")
  }
}

class NfcQuickCheckAsciiTest extends Test, Case {
  override predicate run(Qnit test) {
    if hasNfcNormalizationCheckResult("Hello", -1) = DefinitelyNormalized()
    then test.pass("ASCII string is definitely normalized.")
    else test.fail("ASCII string is not definitely normalized.")
  }
}

class NfcQuickCheckNonNormalizedTest extends Test, Case {
  override predicate run(Qnit test) {
    if hasNfcNormalizationCheckResult(unescapeUnicode("e\\u0340"), _) = DefinitelyNotNormalized()
    then test.pass("String with decomposed character is definitely not normalized.")
    else
      test.fail("String with decomposed character is not recognized as definitely not normalized.")
  }
}

class NfcQuickCheckNormalizedTest extends Test, Case {
  override predicate run(Qnit test) {
    if hasNfcNormalizationCheckResult("é", -1) = DefinitelyNormalized()
    then test.pass("String with precomposed character is definitely normalized.")
    else test.fail("String with precomposed character is not recognized as definitely normalized.")
  }
}

class NfcQuickCheckDefinitelyNotNormalizedIndexTest extends Test, Case {
  override predicate run(Qnit test) {
    if
      forex(int index |
        hasNfcNormalizationCheckResult(unescapeUnicode("e\\u0340 and \\u0341"), index) =
          DefinitelyNotNormalized()
      |
        index = 1 or index = 7
      )
    then test.pass("Index of definitely not normalized character identified correctly.")
    else test.fail("Index of definitely not normalized character not identified correctly.")
  }
}

class NfcQuickCheckMaybeNormalizedTest extends Test, Case {
  override predicate run(Qnit test) {
    if hasNfcNormalizationCheckResult(unescapeUnicode("a\\u0301"), _) = MaybeNormalized()
    then test.pass("String with maybe normalized character identified correctly.")
    else test.fail("String with maybe normalized character not identified correctly.")
  }
}

class NfcQuickCheckMixedMaybeAndNotNormalizedTest extends Test, Case {
  override predicate run(Qnit test) {
    if
      forex(int index, NormalizationCheckResult rslt |
        hasNfcNormalizationCheckResult(unescapeUnicode("a\\u0301 and \\u0340"), index) = rslt
      |
        index = 7 and rslt = DefinitelyNotNormalized()
      )
    then
      test.pass("Index of definitely not normalized character in mixed string identified correctly.")
    else
      test.fail("Index of definitely not normalized character in mixed string not identified correctly.")
  }
}

class NfcQuickCheckMultipleMaybeNormalizedIndicesTest extends Test, Case {
  override predicate run(Qnit test) {
    if
      forex(int index |
        hasNfcNormalizationCheckResult(unescapeUnicode("a\\u0301 and \\u0300"), index) =
          MaybeNormalized()
      |
        index = 1 or index = 7
      )
    then test.pass("Indices of maybe normalized characters identified correctly.")
    else test.fail("Indices of maybe normalized characters not identified correctly.")
  }
}
