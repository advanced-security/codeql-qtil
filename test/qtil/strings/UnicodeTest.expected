| All 21 tests passed. |
